### PR TITLE
Implement more realistic doctests.

### PR DESCRIPTION
**Pull request by jsha**
Add `is_test` and `fn main` headers to various doctests, and use real
URLs along with the `?` operator.

### CODE REVIEW

File: src/agent.rs [hunk 1]
```diff
@@ -274,10 +285,15 @@ impl AgentBuilder {
     /// The default is 30 seconds.
     ///
     /// ```
+    /// # use std::time::Duration;
```
Comment (algesten) [inline src/agent.rs:288]: I think we should show this import.

File: src/agent.rs [hunk 2]
```diff
@@ -288,13 +304,16 @@ impl AgentBuilder {
     /// If both this and `.timeout()` are both set, `.timeout()`
     /// takes precedence.
     ///
-    /// The default is `0`, which means it can block forever.
-    ///
```
Comment (algesten) [inline src/agent.rs:292]: The doc was wrong, but we should say what the default behavior is.

File: src/agent.rs [hunk 3]
```diff
@@ -305,13 +324,16 @@ impl AgentBuilder {
...
```
Comment (algesten) [inline src/agent.rs:309]: Ditto

File: src/agent.rs [hunk 4]
```diff
@@ -346,12 +372,15 @@ impl AgentBuilder {
     /// If the redirect count hits this limit (and it's > 0), TooManyRedirects is returned.
     ///
     /// ```
-    /// let r = ureq::builder()
-    ///     .redirects(10)
```
Comment (algesten) [inline src/agent.rs:350]: Can we use redirects(0) to turn off redirects completely?

File: src/agent.rs [hunk 5]
```diff
@@ -362,11 +391,15 @@ impl AgentBuilder {
     ///
     /// Example:
     /// ```
-    /// let tls_config = std::sync::Arc::new(rustls::ClientConfig::new());
+    /// # fn main() -> Result<(), ureq::Error> {
+    /// # ureq::is_test(true);
+    /// # use std::sync::Arc;
```
Comment (algesten) [inline src/agent.rs:396]: Show this import

File: src/agent.rs [hunk 6]
```diff
@@ -305,13 +324,16 @@ impl AgentBuilder {
     /// If both this and `.timeout()` are both set, `.timeout()`
     /// takes precedence.
     ///
-    /// The default is `0`, which means it can block forever.
-    ///
     /// ```
+    /// # use std::time::Duration;
```
Comment (algesten) [inline src/agent.rs:328]: Show this import

File: src/agent.rs [hunk 7]
```diff
@@ -288,13 +304,16 @@ impl AgentBuilder {
...
```
Comment (algesten) [inline src/agent.rs:308]: Show this import

File: src/agent.rs [hunk 8]
```diff
@@ -381,22 +414,24 @@ impl AgentBuilder {
...
-    /// use cookie_store::CookieStore;
-    /// use std::fs::File;
-    /// use std::io::BufReader;
-    ///
-    /// let file = File::open("cookies.json").unwrap();
+    /// ```
+    /// # fn main() -> Result<(), ureq::Error> {
+    /// # ureq::is_test(true);
+    /// # use cookie_store::CookieStore;
+    /// # use std::fs::File;
```
Comment (algesten) [inline src/agent.rs:421]: Show these imports

File: src/request.rs [hunk 9]
```diff
@@ -62,13 +64,16 @@ impl Request {
     /// Use `.timeout_connect()` and `.timeout_read()` to avoid blocking forever.
     ///
     /// ```
-    /// let r = ureq::builder()
-    ///     .timeout_connect(std::time::Duration::from_secs(10)) // max 10 seconds
+    /// # use std::time::Duration;
```
Comment (algesten) [inline src/request.rs:67]: show

File: src/request.rs [hunk 10]
```diff
@@ -194,13 +208,14 @@ impl Request {
...
     ///
     /// ```
-    /// use std::io::Cursor;
-    ///
-    /// let read = Cursor::new(vec![0x20; 100_000]);
-    ///
-    /// let resp = ureq::post("http://localhost/example-upload")
-    ///     .set("Content-Type", "text/plain")
-    ///     .send(read);
+    /// # use std::io::Cursor;
```
Comment (algesten) [inline src/request.rs:211]: show

File: src/response.rs [hunk 11]
```diff
@@ -213,22 +227,23 @@ impl Response {
     /// Example:
     ///
     /// ```
-    /// # #[cfg(feature = "tls")] {
-    /// use std::io::Read;
-    ///
-    /// let resp =
-    ///     ureq::get("https://ureq.s3.eu-central-1.amazonaws.com/hello_world.json")
-    ///         .call().unwrap();
+    /// # use std::io::Read;
```
Comment (algesten) [inline src/response.rs:230]: show

**Review by algesten — APPROVED** [review]
Very good!✨

Overall I don't think we should hide imports of std stuff. I want ureq to be accessible also for beginners who don't know what a BufRead or a Duration is.

**Comment by jsha** [comment]
Good call on showing the imports. Do you have thoughts on showing them as a separate line (good style) vs using fully-qualified names inline (shorter examples)?

**Comment by algesten** [comment]
I think it's cleaner to have them on separate lines separated by one blank line to the actual example. In some respects I want the examples to be copy-paste into a user's code and I think one should use imports (although I'm a bit lax on it myself).

**Pull request merged**